aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -1658,101 +1666,120 @@ Tensor _matmul_impl(
-    // optimization: transpose the inner dimensions of the arguments, call
-    // matmul on the swapped arguments, then transpose the inner dimensions
-    // of the result.
-    const int64_t n = dim_tensor1 == 2 ? tensor1.size(-2) : 1;
-    const int64_t m = tensor1.size(-1);
-    const int64_t p = tensor2.size(-1);
-
-    const Tensor t2_T = tensor2.transpose(-1, -2);
-    const Tensor t1_T = dim_tensor1 == 2 ? tensor1.t() : tensor1.reshape({n, m}).t();
-    const Tensor res_T = _matmul_impl(out, t2_T, t1_T);
-
-    if (dim_tensor1 == 2) {
-      Tensor res = res_T.transpose(-1, -2).contiguous();
-      return has_out ? out.set_(res) : res;
+    const auto sizes_1 = t1->sizes();
+    auto output_shape = DimVector(sizes_1.begin(), sizes_1.end() - 1);
+    const auto batch_product = c10::multiply_integers(output_shape);
+
+    // Readjust output_shape if we are multiplying by a matrix
+    const auto t2_is_matrix = t2->dim() == 2;
+    if (t2_is_matrix) {
+      output_shape.push_back(t2->sizes()[1]);
     }
-    else {
-      c10::IntArrayRef shape_array = tensor2.sizes().slice(0, dim_tensor2 - 2);
-      DimVector shape(shape_array.begin(), shape_array.end());
-      shape.push_back(p);
+    const auto t1_folded = t1->reshape({batch_product, sizes_1.back()});
+    if (!has_out) {
+      if (t2_is_matrix) {
+        // FIXME This path always does an unnecessary copy when transpose == true as the returned
+        // result from BLAS is already C-transposed
+        const auto output = at::_unsafe_view(t1_folded.mm(*t2), output_shape);
+        return transpose ? output.transpose_(-2, -1).contiguous() : output;
+      } else {
+        return at::_unsafe_view(t1_folded.mv(*t2), output_shape);
+      }
+    } else {
+      // Resize output into the correct shape
+      const auto transpose_out = transpose && t2_is_matrix;
+      if (transpose_out) {
+        // Swap last two elements of output_shape
+        std::iter_swap(output_shape.end() - 2, output_shape.end() - 1);
+        at::native::resize_output(out, output_shape);
+        out.transpose_(-2, -1);
+        std::iter_swap(output_shape.end() - 2, output_shape.end() - 1);
+      } else {
+        at::native::resize_output(out, output_shape);
+      }
 
-      Tensor res = res_T.reshape(shape).contiguous();
-      return has_out ? out.set_(res) : res;
+      // We then reshape the output to the expected shape and call mm/mv
+      // and transpose back if necessary
+      auto reshaped_out = t2_is_matrix ? out.reshape({batch_product, t2->sizes().back()})
+                                       : out.reshape({batch_product});
+      if (t2_is_matrix) {
+        at::mm_out(reshaped_out, t1_folded, *t2);
+      } else {
+        at::mv_out(reshaped_out, t1_folded, *t2);
+      }
+      if (!reshaped_out.is_alias_of(out)) {
+        out.copy_(reshaped_out.view_as(out));
+      }
+      if (transpose_out) {
+        out.transpose_(-2, -1);
```

Comment:
You untranspose it later, but the damage is done: another thread reading the metadata for this tensor could temporarily see the "altered" metadata here; and there wouldn't be expected to be a race if the other thread was only supposed to be writing into the data (which is what `out=` is supposed to do)

Reply:
Or if you want to avoid it when the input strides match exactly, a reference to a possibly allocated tensor

Reply:
I'll push the changes in a second.

I did the change and tested that it is indeed working as expected since the tests pass, and the tests fail if you put a `TORCH_CHECK(!transpose_out);` inside the `if (!reshaped_out.is_alias_of(out)) {` if, the condition fires.